### PR TITLE
Clean-up error checking in linalg.

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80767

This PR does a number of things:
- Fixes https://github.com/pytorch/pytorch/issues/80735 returning early in the most common case (no errors).
- Leaves just one entry to check the errors, so that it's as uniform as it gets (easy to get right, difficult to get wrong)
- Having all the error checking in one function simplifies the code
- We make sure that the `infos_ptr` sent over to LAPACK is always of type `int`, as it should
- We just access `infos.data_ptr<int32_t>` in `_linalg_check_errors`

Edit. I get a consistent x7-10 speed-up with this patch using the script in https://github.com/pytorch/pytorch/issues/80735, so I think we're good here.

### CODE REVIEW

**Comment by Lezcano** [comment]
@IvanYashchuk PTAL

File: aten/src/ATen/native/LinearAlgebraUtils.h
Comment (IvanYashchuk) [inline aten/src/ATen/native/LinearAlgebraUtils.h:302]: What's the benefit of moving the code from here to `aten/src/ATen/native/BatchLinearAlgebra.cpp` instead of introducing the changes in this file?

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp
Comment (Lezcano) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp:2355]: The day we're able to merge those PRs and remove all that legacy code I'll ship a bottle of champagne to Helsinki.

File: aten/src/ATen/native/LinearAlgebraUtils.h
Comment (Lezcano) [reply aten/src/ATen/native/LinearAlgebraUtils.h:302]: Now there's just one entry point to this function (through `_linalg_check_errors`) so that way it's impossible to use wrong and we always get something that's composite compliant and all that.

**Comment by Lezcano** [comment]
@IvanYashchuk addressed the review PTAL.

**Comment by ngimel** [comment]
@IvanYashchuk is this PR good to go?

**Review by IvanYashchuk — APPROVED** [review]
Good to go. Thank you, Mario!

**Comment by Lezcano** [comment]
@janeyx99 I believe that I should be able to merge this after Ivan approved it, but the merge has failed. Could you check what's going on there?

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**
